### PR TITLE
Fix phpstan field options

### PR DESCRIPTION
**Pull request by franmomu**
Follow-up of https://github.com/sonata-project/SonataAdminBundle/pull/7099#pullrequestreview-640872966 to merge to master is not affected.

### CODE REVIEW

File: src/FieldDescription/FieldDescriptionInterface.php [hunk 1]
```diff
@@ -53,7 +53,7 @@
  *  translation_domain?: string,
  *  type?: string,
  *  virtual_field?: bool
- * }&array<string, mixed>|array<empty, empty>
+ * }|array<empty, empty>
```
Comment (jordisala1991) [inline src/FieldDescription/FieldDescriptionInterface.php:56]: If we don't remove this one it fails ?

**Comment by VincentLanglet** [comment]
Copy of my last response in https://github.com/sonata-project/SonataAdminBundle/pull/7099

> > Restricting the option in the interface FieldDescriptionInterface, it's like restricting the option in the FormTypeInterface.
> > The options resolver is specific to a FormType. Each FormType has custom options.
> > It's the same here: each FieldDescription could have custom option.
> 
> I think it's quite different, by general, there is only one or two implementations (one per persistence bundle) of `FieldDescriptionInterface`, you don't create an implementation of FieldDescriptionInterface for a custom field.

Yes, but someone could use his own persistence or his own FieldDescription with more features (and options).
This should be allowed.

> The options, as stated [in the code](https://github.com/sonata-project/SonataAdminBundle/blob/4b1a41dab94ab956d99090110ad2f560cc529a1c/src/FieldDescription/BaseFieldDescription.php#L32-L64), are dependent of the type of the context.

Some options are specific to only some FieldDescription type ; for instance `timezone`, which is only used for `date` values.
https://github.com/sonata-project/SonataAdminBundle/blob/4b1a41dab94ab956d99090110ad2f560cc529a1c/src/FieldDescription/FieldDescriptionInterface.php#L52

or `currency` https://github.com/sonata-project/SonataAdminBundle/blob/2c07756867f2a1a466c5292483cd05e50a1932ff/src/Resources/views/CRUD/display_currency.html.twig#L16

which only related to the `TYPE_CURRENCY` and is not in the `FieldDescriptionOptions` currently.

another example `inverse`: https://github.com/sonata-project/SonataAdminBundle/blob/2c07756867f2a1a466c5292483cd05e50a1932ff/src/Resources/views/CRUD/display_boolean.html.twig#L19

`format`: https://github.com/sonata-project/SonataAdminBundle/blob/2c07756867f2a1a466c5292483cd05e50a1932ff/src/Resources/views/CRUD/display_time.html.twig#L17

I'm sure I can find others options used for specific FieldDescription types.
And since we're allowing custom types, we shouldn't restrict the options for them.

> > And when using a custom template, options are passed to the template, so it's a way to pass special values to the view, so basically any value could be possible.
> 
> For passing values to the template there should be an option for that, to use options that way it's just a hack and I think there is no documentation about that.

I just follow the way we're doing for all the existing templates.

File: src/FieldDescription/FieldDescriptionInterface.php (hunk 1 above)
Comment (franmomu) [reply src/FieldDescription/FieldDescriptionInterface.php:56]: yes

**Comment by jordisala1991** [comment]
How do we move forward? This is blocking the master branch

**Comment by VincentLanglet** [comment]
> How do we move forward? This is blocking the master branch

As I said, for me the `FieldDescriptionOptions` is wrong and should be removed in favor of `array<string, mixed>` since we're using a lot more options than the one in the FieldDescriptionOptions and a user could have some specific options for custom types.

What If I wanted to create my own type with a `precision` options ?

**Comment by franmomu** [comment]
what kind of type?

**Comment by VincentLanglet** [comment]
> what kind of type?
> 
> We are trying to close API and restrict things, if we need options I can add them, but not leave this open to anything.

If we look
https://symfony.com/doc/current/bundles/SonataAdminBundle/reference/field_types.html#list-and-show-actions

There is at least `format`, `timezone`, `as_string`, `subject`, `body`, `catalogue`, `currency`, `editable`, `inline`, `display`, `key_translation_domain`, `value_translation_domain`, `ajax_hidden`, `inverse`, `choices`, `multiple`, `delimiter`, `class`, `required`, `hide_protocol`, `url`, `attributes`, `route.name`, `route.parameters`,  `route.absolute`, `route.identifier_parameter_name`, `strip`, `truncate`, `truncate.length`, `truncate.cut`, `truncate.ellipsis`.

I agree we're trying to close API, but the field description is something open by pattern since custom type are allowed:
https://symfony.com/doc/current/bundles/SonataAdminBundle/reference/field_types.html#create-your-own-field-type
And I see no reason to forbid the developer to pass options to his custom type the same way we're doing it.

I could create my own type `amount`
```
->add('initialAmount', 'amount', [
    'currency' => true, // A custom option to display or not the currency
    'precision' => 2, // A custom option for the number of digits after the coma.
])
```
Here, the developper will have two issues: 
- `currency` will be restricted to `string` in `FieldDescriptionsOption` because we use it as a string in the `FieldDescriptionInterface::TYPE_CURRENCY`, but the `amount` type has nothing in common and works with a `boolean` I shouldn't have to rename the option just because it's used by another fieldDescription type.
- `precision` doesn't exist in `FieldDescriptionsOption` if we don't add `&array<string, mixed>` but the code written is valid.

The issue here is that we have generic options but also specific (to only some field description types) ones and we pass them the same way. If something should be in the FieldDescriptionsOption, it should only be the generic options for all the possible types. Then we add `&array<string, mixed>` to allow any custom specific options.

**Comment by franmomu** [comment]
To me the right approach would be to do as we do with form that has a separated options:

https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Form/FormMapper.php#L80

and filters:

https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Datagrid/DatagridMapper.php#L78-L85

**Comment by jordisala1991** [comment]
> To me the right approach would be to do as we do with form that has a separated options:
> 
> https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Form/FormMapper.php#L80
> 
> and filters:
> 
> https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Datagrid/DatagridMapper.php#L78-L85

That's something worth considering, but for now, the array should accept any options, like it was before. And an issue could be opened for tackle that problem.

**Comment by VincentLanglet** [comment]
> To me the right approach would be to do as we do with form that has a separated options.

It could something to start with, keeping in mind that 
- For the form, the field descriptions options are added to the options https://github.com/sonata-project/SonataAdminBundle/blob/3.x/src/Builder/AbstractFormContractor.php#L70 before being passed to the form builder https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Form/FormMapper.php#L196
- For the filters, they are merged https://github.com/sonata-project/SonataAdminBundle/blob/a8731dc05c4546d83149576ee446dc1cc1daf7db/src/Datagrid/DatagridMapper.php#L137

Personally, when using these methods, I never used the second array of option and put all my options directly in the first array  because I never understood which options should be in the first array, and which should be in the second one.

But we should indeed
- either use two options params, which shouldn't be merged like we do actually, one for generic options, one for custom options.
- either moved all the custom options to a new key like `template_options`.

This seems to need a lot of refacto, BC-break, reflexions and discussions so I agree with jordisala, the quickest way to solve the issue is to remove `FieldDescriptionsOptions` in favor of `array<string, mixed>` and open an issue.

**Pull request closed without merging**